specs/networking/rpc-interface.md

```diff
@@ -0,0 +1,246 @@
+5. The `error` member is REQUIRED on errors, and MUST NOT exist if there wasn't an error.
+
+Structuring RPC requests in this manner allows multiple calls and responses to be multiplexed over the same stream without switching.
+
+The "method ID" fields in the below messages refer to the `method` field in the request structure above.
+
+The first 1,000 values in `error.code` are reserved for system use. The following error codes are predefined:
+
+1. `0`: Parse error.
+2. `10`: Invalid request.
+3. `20`: Method not found.
+4. `30`: Server error.
+
+## Messages
+
+### Hello
+
+**Method ID:** `0`
+
+**Body**:
+
+```
+(
+    network_id: uint8
+    latest_finalized_root: bytes32
+    latest_finalized_epoch: uint64
+    best_root: bytes32
+    best_slot: uint64
+)
+```
+
+Clients exchange `hello` messages upon connection, forming a two-phase handshake. The first message the initiating client sends MUST be the `hello` message. In response, the receiving client MUST respond with its own `hello` message.
+
+Clients SHOULD immediately disconnect from one another following the handshake above under the following conditions:
+
+1. If `network_id` belongs to a different chain, since the client definitionally cannot sync with this client.
+2. If the `latest_finalized_root` shared by the peer is not in the client's chain at the expected epoch. For example, if Peer 1 in the diagram below has `(root, epoch)` of `(A, 5)` and Peer 2 has `(B, 3)`, Peer 1 would disconnect because it knows that `B` is not the root in their chain at epoch 3:
+
+```
+              Root A
+
+              +---+
+              |xxx|  +----+ Epoch 5
+              +-+-+
+                ^
+                |
+              +-+-+
+              |   |  +----+ Epoch 4
+              +-+-+
+Root B          ^
+                |
++---+         +-+-+
+|xxx+<---+--->+   |  +----+ Epoch 3
++---+    |    +---+
+         |
+       +-+-+
+       |   |  +-----------+ Epoch 2
+       +-+-+
+         ^
+         |
+       +-+-+
+       |   |  +-----------+ Epoch 1
+       +---+
+```
+
+Once the handshake completes, the client with the higher `latest_finalized_epoch` or `best_slot` (if the clients have equal `latest_finalized_epoch`s) SHOULD send beacon block roots to its counterparty via `beacon_block_roots` (i.e., RPC method `10`).
+
+### Goodbye
+
+**Method ID:** `1`
+
+**Body:**
+
+```
+(
+    reason: uint64
+)
+```
+
+Client MAY send `goodbye` messages upon disconnection. The reason field MUST be one of the following values:
+
+- `1`: Client shut down.
+- `2`: Irrelevant network.
+- `3`: Irrelevant shard.
+
+### Provide Beacon Block Roots
+
+**Method ID:** `10`
+
+**Body:**
+
+```
+# BlockRootSlot
+(
+    block_root: HashTreeRoot
+    slot: uint64
+)
+
+(
+    roots: []BlockRootSlot
+)
+```
+
+Send a list of block roots and slots to the peer.
+
+### Beacon Block Headers
+
+**Method ID:** `11`
+
+**Request Body**
+
+```
+(
+    start_root: HashTreeRoot
```

Comment:
Do we need the root? It seems redundant to me, except for the case of chain reorgs which shouldn't happen frequently at sync (and even then, it's probably better to get blocks from the current chain that we'll be able to use later, instead of getting outdated ones).

Reply:
ẁe need a mechanism for recovering blocks, in case something is lost or the client goes offline for a short bit and loses a few (computer went to sleep / ISP went down for 10 minutes). 

I argue in the original issue (https://github.com/ethereum/eth2.0-specs/issues/692#issuecomment-469356620) that  it's often natural to request blocks backwards for this reason: the data structure we're syncing is a singly linked list pointing backwards in time and we receive attestations and blocks that let us discover heads "naturally" by listening to the broadcasts. With a `block_root+previous_n_blocks` kind of request we can both sync and recover, and for example use attestations to discover "viable" heads to work on, from a sync or recovery perspective. Indeed, negotiating finalized epochs in the handshake is somewhat redundant in that case, albeit a nice optimization (except for the chain id) - we could equally well request blocks from the peer that gossiped us the block or attestation whose parent we're missing - they should not be gossiping attestations they have not linked to a finalized epoch of value.

Reply:
Interesting! To summarize my understanding of your comment: Syncing forwards is safer as we can verify each block immediately when we receive it, but syncing backwards is more efficient/doesn't require additional database indexing (and I guess syncing forwards may require a negotiating phase to discover the best shared block). You're proposing to interpret the fact that I see lots of attestations on top of my sync peer's head flying around the network as evidence that their head is valid? And therefore, I'd be pretty safe syncing backwards?

That sounds reasonable. My original concern was that this requires me to know (at least some good fraction of) the validator set as otherwise my sync peer could create lots of fraudulent attestations for free that I have no chance of verifying. But I would notice this if I have at least one single honest peer (if I try to sync from them or compare the attestations coming from them).

Do you think having only a backwards sync is fine or do we need both (e.g. for highly adversarial environments, or resource constrained devices that don't participate in gossiping?).

Reply:
> more efficient

In terms of network / bandwidth, I'd say it's about the same but there are some nuances:

* in forward sync, I can ask for "more" slots than already exist, potentially saving round trips - a client could use this to request "all" data at the time of request arrival. consider the following race: A sends request, a new block is produced, B receives request (similar: B starts sending response which takes time, new block is produced).
* in backward sync, one could have an `(latest-head, known_slot_number)` request ("give me the block you consider to be the head, and history back to slot N") to alleviate this race, but then the server selects the head.
* both above races are generally solved by collecting data from broadcasts while syncing (classic subscribe-then-sync pattern) - they are mainly concerns if you don't want to subscribe or want to delay subscribing.
* in forward sync, I might end up on a different branch / head than I thought I would - the request itself does not point one out

In terms of client implementations, I think of backward sync as biased to make it cheaper for the server: the server already has the data necessary - also because the head is kept hot - while the client has to keep a chain of "unknown" blocks around / can't validate eagerly. An additional rule that the response must be forward-ordered could help the client apply / validate the blocks eagerly.

The backwards sync can be seen as more passive/reactive/lazy while forward sync is more active..

> attestations on top of my sync peer's head flying around the network as evidence that their head is valid

right. the assumption rests on two premises (thanks @djrtwo!):

* honest clients will not propagate invalid data (not signed by a validator they know)
* there's a slashing condition on creating unviable attestations - there's currently no penalty to create & sign an unviable block so one can perhaps imagine a malicious group of validators creating lots of these and for example spam everyone during important periods "for free". It sounds a bit far fetched though, tbh, to be creating blocks this way - would love to hear thoughts.
* I've weak-subjectively selected an initial state that contains some validators. I'd primarily look for anything signed by those validators as another heuristic for where to start syncing (even if the validator set might have changed from there).

> Do you think having only a backwards sync is fine or do we need both (e.g. for highly adversarial environments, or resource constrained devices that don't participate in gossiping?).

I'm not sure :) I'm curious to hear feedback on this point, but here are some thoughts:
* it's important that we have a request like Hello to ask for what clients consider to be the head for non-gossiping use cases - but I think that's orthogonal to the sync direction.
* clients should be free to send that request at any time, not just during the initial negotiation phase
* direction can be different for request and response - if different, requires a slightly "smarter" server
* there's a cost for each direction, in terms of implementation. I'd start with one and look for strong motivations before implementing the other, as the returns are not that great. Either direction is sufficient, really.

Reply:
> Do you think having only a backwards sync is fine or do we need both (e.g. for highly adversarial environments, or resource constrained devices that don't participate in gossiping?).

It seems reasonable to sync backwards from the latest received gossiped block (at least as an initial implementation)

> in backward sync, one could have an (latest-head, known_slot_number) request ("give me the block you consider to be the head, and history back to slot N") to alleviate this race, but then the server selects the head.

i would argue that we should drop start_slot.  if we give clients the option to request a block by either start_slot or start_root then that forces us to maintain a lookup or search mechanism for both.  if we are saying that both fields (start_slot and start_root) required to sync, then I would disagree.  we should be able to simply perform a lookup by block_root and walk the chain backwards until we reach max_headers.

Reply:
>  latest received gossiped block

or even better, latest gossiped attestation

> Do we really need start_slot?

I would say that if we go with backwards sync, we should not implement forwards sync here or elsewhere unless there's a strong case for that direction. Having to implement both directions negates some of the benefits of backward sync and adds implementation surface.

It is quite possible to add forward sync in a later version of the protocol as well should it prove necessary.

Reply:
> or even better, latest gossiped attestation

I can dig that

@arnetheduck or anyone else. Why do we need start_slot